chainstate/src/detail/chainstateref.rs

```diff
@@ -374,7 +374,62 @@ impl<'a, S: BlockchainStorageRead, O: OrphanBlocks> ChainstateRef<'a, S, O> {
         Ok(())
     }
 
-    fn check_block_detail(&self, block: &Block) -> Result<(), CheckBlockError> {
+    fn check_block_reward_maturity_settings(&self, block: &Block) -> Result<(), CheckBlockError> {
+        // TODO: test every individual case
+        let required = block.consensus_data().reward_maturity_distance(self.chain_config);
+        for output in block.block_reward().outputs() {
+            match output.purpose() {
+                common::chain::OutputPurpose::Transfer(_) => {
+                    return Err(CheckBlockError::InvalidBlockRewardOutputType(
+                        block.get_id(),
+                    ))
+                }
+                common::chain::OutputPurpose::LockThenTransfer(_, tl) => match tl {
+                    common::chain::timelock::OutputTimeLock::UntilHeight(_) => {
+                        return Err(CheckBlockError::InvalidBlockRewardMaturityTimelockType(
+                            block.get_id(),
+                        ))
+                    }
+                    common::chain::timelock::OutputTimeLock::UntilTime(_) => {
+                        return Err(CheckBlockError::InvalidBlockRewardMaturityTimelockType(
+                            block.get_id(),
+                        ))
+                    }
+                    common::chain::timelock::OutputTimeLock::ForBlockCount(c) => {
+                        let cs: i64 = (*c).try_into().map_err(|_| {
+                            CheckBlockError::InvalidBlockRewardMaturityDistanceValue(
+                                block.get_id(),
+                                *c,
+                            )
+                        })?;
+                        let given = BlockDistance::new(cs);
```

Comment:
Not really related to this PR, but `OutputTimeLock::ForBlockCount` should really take `NonNegativeBlockDistance` or something like that rather than `u64`. The conversions here are a bit awkward.

Reply:
Making it nice is something we can do later. But having it signed is an issue because we need a compact form for it, and signed doesn't support compact.